src/test/java/com/amazon/dlic/auth/http/jwt/keybyoidc/SelfRefreshingKeySetTest.java

```diff
@@ -11,65 +11,98 @@
+
     @Test
-    public void basicTest() throws AuthenticatorUnavailableException, BadCredentialsException {
-        SelfRefreshingKeySet selfRefreshingKeySet = new SelfRefreshingKeySet(new MockKeySetProvider());
-
-        OctetSequenceKey key1 = (OctetSequenceKey) selfRefreshingKeySet.getKey("kid/a");
-        Assert.assertEquals(TestJwk.OCT_1_K, key1.getKeyValue().decodeToString());
-        Assert.assertEquals(1, selfRefreshingKeySet.getRefreshCount());
-
-        OctetSequenceKey key2 = (OctetSequenceKey) selfRefreshingKeySet.getKey("kid/b");
-        Assert.assertEquals(TestJwk.OCT_2_K, key2.getKeyValue().decodeToString());
-        Assert.assertEquals(1, selfRefreshingKeySet.getRefreshCount());
-
-        try {
-            selfRefreshingKeySet.getKey("kid/X");
-            Assert.fail("Expected a BadCredentialsException");
-        } catch (BadCredentialsException e) {
-            Assert.assertEquals(2, selfRefreshingKeySet.getRefreshCount());
-        }
+    public void getKey_withKidShouldReturnValidKey() throws AuthenticatorUnavailableException, BadCredentialsException {
 
+        OctetSequenceKey key = (OctetSequenceKey) selfRefreshingKeySet.getKey("kid/a");
+        assertThat(TestJwk.OCT_1_K, is(equalTo(key.getKeyValue().decodeToString())));
     }
 
-    @Test(timeout = 10000)
-    public void twoThreadedTest() throws Exception {
-        BlockingMockKeySetProvider provider = new BlockingMockKeySetProvider();
+    @Test
+    public void getKey_withNullKidShouldThrowAuthenticatorUnavailableException() throws AuthenticatorUnavailableException,
+        BadCredentialsException {
 
-        final SelfRefreshingKeySet selfRefreshingKeySet = new SelfRefreshingKeySet(provider);
+        Assert.assertThrows(AuthenticatorUnavailableException.class, () -> selfRefreshingKeySet.getKey(null));
 
-        ExecutorService executorService = Executors.newCachedThreadPool();
+    }
 
-        Future<JWK> f1 = executorService.submit(() -> selfRefreshingKeySet.getKey("kid/a"));
+    @Test
+    public void getKey_withInvalidDataShouldReturnBadCredentialException() throws AuthenticatorUnavailableException,
+        BadCredentialsException {
 
-        provider.waitForCalled();
+        Assert.assertThrows(BadCredentialsException.class, () -> selfRefreshingKeySet.getKey("kid/X"));
+    }
 
-        Future<JWK> f2 = executorService.submit(() -> selfRefreshingKeySet.getKey("kid/b"));
+    @Test
+    public void getKeyAfterRefresh_withKidShouldReturnKey() throws AuthenticatorUnavailableException, BadCredentialsException {
+
+        OctetSequenceKey key = (OctetSequenceKey) selfRefreshingKeySet.getKeyAfterRefresh("kid/b");
+        assertThat(TestJwk.OCT_2_K, is(equalTo(key.getKeyValue().decodeToString())));
+    }
 
-        while (selfRefreshingKeySet.getQueuedGetCount() == 0) {
-            Thread.sleep(10);
+    @Test
+    public void getKeyAfterRefresh_queuedGetCountVariableShouldBeZeroWhenFinishWithAllKeyRefreshes() throws InterruptedException,
```

Comment:
Looks like this test was one of the ones that's reporting failures based on how fast the threads process in the GHA vs a local machine.

What I'd recommend is create a SelfRefreshingKeySet with your own KeySetProvider that you can control how quickly before there is a response.  This way, you can have it sit and wait, make some test assertions, then have it return a key when you are ready and test different assertions when it should be done.  This will remove the 'random' behavior because you'll know exactly when the KeySetProvider is refreshing vs queuing vs completed.

Reply:
Another thing is, this failed could prove that the code is not executed synchronously as I mentioned in this [comment](https://github.com/opensearch-project/security/pull/3509#discussion_r1383685122); If so, it would have been 10 when it reached the assertion, am I wrong?

Reply:
Currently the tests use `MockKeySetProvider`, which immediately returns a JWK.  This is not very useful for this async testing - there is no time for anything interesting to happen, instead create a new class like the following that you can control if it waits or doesn't wait.  By using a countdownlatch you can control when to allow execution to continue or stop.  There are more advanced concurrency privatives that you can use for more advanced conditions, I'd recommend starting with [[java.util.concurrency.CountDownLatch]](https://docs.oracle.com/javase/8/docs/api/index.html?java/util/concurrent/CountDownLatch.html).

```java
public class LatchedTestKeySetProvider implements KeySetProvider {

    private final CountDownLatch latch;
    public LatchTestKeySetProvider(CountDownLatch latch) {
        this.latch = latch;
    }

    @Override
    public JWKSet get() throws AuthenticatorUnavailableException {
        try {
            // Wait until the latch is counted down
            latch.await();
            return TestJwk.OCT_1_2_3
        } catch (final InterruptedException e) {
            throw new AuthenticatorUnavailableException("Thread interrupted", e);
        }
    }
}
```

Reply:
Perfect, I'll try it. Thanks for the guidance.

Reply:
This is why I learned working with CountDownLatch.
@peternied I created the class using the idea you suggested, then I had the problem that the `latch.await()` didn't make the main Thread to wait, so I got random responses every time I run the test.
Then I found out that there is another thread created inside the synchronized method `private synchronized JWK performRefresh(String kid) {` so, my threads (using latch) didn't wait for that thread to finish and that was the root cause of my random responses, since sometimes this new thread was slow, and sometimes was fast, so my assertions didn't work for the answers I was looking for.
I decided to create a test that just make a lot of asynchronous calls to the method, so the refreshCount and the queueCount get increased. I couldn't make it work with specific numbers because of the random responses I got, so I went for greaterThan 0.
I hope I explained myself, if not, please let me know.

Reply:
Also, when running tests I got randomly errors for this line where `queuedGetCount` is always 0 even thought you use CountDownLatch.
https://github.com/opensearch-project/security/pull/3509/files#diff-e852cf3e1cb5c1b2226df1b2b3cea2a06ba861823c634c52ebd074ffd41cd727L62

![image](https://github.com/opensearch-project/security/assets/48450162/be2b62fd-e426-4e95-98a2-030318b2e7f1)

Reply:
Did you manage to get to the bottom of this?

Reply:
This line [`Future<?> future = threadPoolExecutor.submit(new Runnable() {`](https://github.com/opensearch-project/security/blob/07833294e750be097f75c8ccd7d40b882147112f/src/main/java/com/amazon/dlic/auth/http/jwt/keybyoidc/SelfRefreshingKeySet.java#L208)  is what makes the tests to get random responses, because the method is synchronized but it creates another thread inside, so it doesn't wait for this new thread to finish its job.

Here is a picture of the random calls made to each thread, and the test failed because sometimes it doesn't reach the point to count before the test ends. As you can see in the picture there is a number `9` when expected to be `10`, and I got a different values every time I run it.  

I removed the line I mentioned, then when I run the test I got good responses., they were the same every time, so I think this is something that needs to be investigated further.

![Screenshot from 2023-12-07 09-08-56](https://github.com/opensearch-project/security/assets/48450162/0e5f0509-0695-4730-b80e-7a336198d746)

Reply:
I'd like to see the majority of this change merged - what do you think about creating a specific follow up issue and removing anything that doesn't behave as you'd expect so we can merge?

Reply:
@davidosorno Please follow up if you feel there is anything else lingering I'm going to resolve this comment and merge.